### ISSUE TITLE
Docker /etc/hosts and DNS is not what we want

### ISSUE THREAD

**Issue by thockin** [kubernetes/kubernetes#14633]
I noticed this earlier and can not explain it:

```
root@kubernetes-minion-gn8a:/home/thockin# cat /etc/resolv.conf 
domain c.thockin-dev.internal.
search c.thockin-dev.internal. 844790599918.google.internal. google.internal.
nameserver 169.254.169.254
nameserver 10.240.0.1
root@kubernetes-minion-gn8a:/home/thockin# nslookup 10.244.2.6 10.0.0.10
Server:         10.0.0.10
Address:        10.0.0.10#53

** server can't find 6.2.244.10.in-addr.arpa.: NXDOMAIN

root@kubernetes-minion-gn8a:/home/thockin# nslookup 10.244.2.6 169.254.169.254
Server:         169.254.169.254
Address:        169.254.169.254#53

** server can't find 6.2.244.10.in-addr.arpa.: NXDOMAIN

root@kubernetes-minion-gn8a:/home/thockin# nslookup 10.244.2.6 10.240.0.1
Server:         10.240.0.1
Address:        10.240.0.1#53

** server can't find 6.2.244.10.in-addr.arpa.: NXDOMAIN
```

but

```
root@kubernetes-minion-gn8a:/home/thockin# docker run -ti busybox
/ # nslookup 10.244.2.6 10.0.0.10
Server:    10.0.0.10
Address 1: 10.0.0.10

Name:      10.244.2.6
Address 1: 10.244.2.6 k8s_POD-7be6d81d_hostnames-ctfjl_default_abdc3cba-58d4-11e5-9447-42010af0f439_85570955
/ # nslookup 10.244.2.6 169.254.169.254
Server:    169.254.169.254
Address 1: 169.254.169.254 metadata.google.internal

Name:      10.244.2.6
Address 1: 10.244.2.6 k8s_POD-7be6d81d_hostnames-ctfjl_default_abdc3cba-58d4-11e5-9447-42010af0f439_85570955
/ # nslookup 10.244.2.6 8.8.8.8
Server:    8.8.8.8
Address 1: 8.8.8.8 google-public-dns-a.google.com

Name:      10.244.2.6
Address 1: 10.244.2.6 k8s_POD-7be6d81d_hostnames-ctfjl_default_abdc3cba-58d4-11e5-9447-42010af0f439_85570955
/ # nslookup 10.244.2.6 127.0.0.1
Server:    127.0.0.1
Address 1: 127.0.0.1 localhost

Name:      10.244.2.6
Address 1: 10.244.2.6 k8s_POD-7be6d81d_hostnames-ctfjl_default_abdc3cba-58d4-11e5-9447-42010af0f439_85570955
```

What is happening here?  Is it docker?  The same doesn't seem to happen in ubuntu.

```
root@kubernetes-minion-gn8a:/home/thockin# docker run -ti ubuntu
root@52401dfed17a:/# apt-get update > /dev/null
root@52401dfed17a:/# yes | apt-get -y install dnsutils > /dev/null
root@52401dfed17a:/# nslookup 10.244.2.6 10.0.0.10
Server:         10.0.0.10
Address:        10.0.0.10#53

** server can't find 6.2.244.10.in-addr.arpa: NXDOMAIN

root@52401dfed17a:/# nslookup 10.244.2.6 169.254.169.254
Server:         169.254.169.254
Address:        169.254.169.254#53

** server can't find 6.2.244.10.in-addr.arpa: NXDOMAIN
```

We need to understand this ASAP.  Docker is version 1.8.2

**Comment by thockin** [comment]
Oh sweet deity, docker is populating /etc/hosts

```
 root@kubernetes-minion-gn8a:/home/thockin# docker run -ti busybox
/ # cat /etc/hosts
10.244.2.14     6396a0f993b0
127.0.0.1       localhost
::1     localhost ip6-localhost ip6-loopback
fe00::0 ip6-localnet
ff00::0 ip6-mcastprefix
ff02::1 ip6-allnodes
ff02::2 ip6-allrouters
10.244.2.2      k8s_POD-7be6d81d_fluentd-cloud-logging-kubernetes-minion-gn8a_kube-system_a55c4c5ce01f12bf433986d82159a3b9_bba18f23.bridge
10.244.2.6      k8s_POD-7be6d81d_hostnames-ctfjl_default_abdc3cba-58d4-11e5-9447-42010af0f439_85570955
10.244.2.14     mad_colden
10.244.2.4      k8s_POD-6e934112_kube-dns-v9-xv0lm_kube-system_f8f49b23-58c6-11e5-9447-42010af0f439_4583e31f.bridge
10.244.2.6      k8s_POD-7be6d81d_hostnames-ctfjl_default_abdc3cba-58d4-11e5-9447-42010af0f439_85570955.bridge
10.244.2.7      k8s_POD-7be6d81d_busybox-rxrev_default_6189dc95-6594-11e5-abfe-42010af0f439_816b884b.bridge
10.244.2.14     mad_colden.bridge
10.244.2.2      k8s_POD-7be6d81d_fluentd-cloud-logging-kubernetes-minion-gn8a_kube-system_a55c4c5ce01f12bf433986d82159a3b9_bba18f23
10.244.2.3      k8s_POD-8d19fd44_kube-registry-proxy-kubernetes-minion-gn8a_kube-system_3119a0eb0b56d0e6997609346df66789_3fa525a3
10.244.2.4      k8s_POD-6e934112_kube-dns-v9-xv0lm_kube-system_f8f49b23-58c6-11e5-9447-42010af0f439_4583e31f
10.244.2.7      k8s_POD-7be6d81d_busybox-rxrev_default_6189dc95-6594-11e5-abfe-42010af0f439_816b884b
10.244.2.3      k8s_POD-8d19fd44_kube-registry-proxy-kubernetes-minion-gn8a_kube-system_3119a0eb0b56d0e6997609346df66789_3fa525a3.bridge
10.244.2.5      k8s_POD-89e0fcaf_kube-registry-v0-bm9ex_kube-system_f8a25492-58c6-11e5-9447-42010af0f439_f4d56a83
10.244.2.5      k8s_POD-89e0fcaf_kube-registry-v0-bm9ex_kube-system_f8a25492-58c6-11e5-9447-42010af0f439_f4d56a83.bridge
```

**Comment by thockin** [comment]
I can't find any docs on this except that Docker does it for links.  I did not expect it for all containers on a machine.  This is not as disastrous as intercepting DNS (phew), but it is really not something we want.

**Comment by thockin** [comment]
https://github.com/docker/docker/issues/16619

**Comment by streamnsight** [comment]
Docker populates the hosts file only with links as far as i know, but it may populate for all containers if the -net=host flag is set.

**Comment by thockin** [comment]
I think this has to do with libnetwork taking over inside docker and
muddling "services" with "networks".  My guess with very little evidence is
that someone figured it made sense to support DNS for containers in the
same Network, but DNS names are tied to container names.  Do Not Want.

**Comment by thockin** [comment]
https://github.com/docker/docker/pull/16645 retools some of this.  We need to figure out if we can get out of the /etc/hosts management they are doing.

**Comment by linzichang** [comment]
@thockin I have ran into a similar sutiation about 2 mouth ago. Docker is version 1.6.0. And docker will populate `/etc/hosts`. In our use case, we want to use our own hosts, will update in the running process and also keep it after container stop or restart. We wanted to test our app ASAP. So we did a really ugly thing to achieve our case: 
1. put the wanted `hosts`(maybe an empty file) in host path `/myapp/hosts`
2. app's run.sh add an operation of rewriting the /etc/hosts with /tmp/hosts
3. docker run -v /myapp/hosts:/tmp/hosts app run.sh

In docker 1.8.2, just need:
1. put the wanted `hosts`(maybe an empty file) in host path `/myapp/hosts`
2. docker run -v /myapp/hosts:/etc/hosts app run.sh

Just a temporary solution in our case. I think in kubernetes we should get rid of the docker's `/etc/hosts` management and just let it contain `localhost` in common case. Is there some cases that kubernetes need to modify the `/etc/hosts` ?
